core/src/repair_service.rs

```diff
@@ -238,19 +286,212 @@ impl RepairService {
+    ) -> Result<()> {
+        let req = serve_repair.map_repair_request(&repair_type, repair_stats)?;
+        repair_socket.send_to(&req, to)?;
+        Ok(())
+    }
+
+    fn update_duplicate_slot_repair_addr(
+        slot: Slot,
+        status: &mut DuplicateSlotRepairStatus,
+        cluster_slots: &ClusterSlots,
+        serve_repair: &ServeRepair,
+    ) {
+        let now = timestamp();
+        if status.repair_addr.is_none()
+            || now.saturating_sub(status.start) >= MAX_DUPLICATE_WAIT_MS as u64
+        {
+            let repair_addr =
+                serve_repair.repair_request_duplicate_compute_best_peer(slot, cluster_slots);
+            status.repair_addr = repair_addr.ok();
+            status.start = timestamp();
+        }
+    }
+
+    fn process_new_duplicate_slots(
+        new_duplicate_slots: &[Slot],
+        duplicate_slot_repair_statuses: &mut HashMap<Slot, DuplicateSlotRepairStatus>,
+        cluster_slots: &ClusterSlots,
+        root_bank: &Bank,
+        blockstore: &Blockstore,
+        serve_repair: &ServeRepair,
+        duplicate_slots_reset_sender: &DuplicateSlotsResetSender,
+    ) {
+        for slot in new_duplicate_slots {
+            warn!(
+                "Cluster completed slot: {}, dumping our current version and repairing",
+                slot
+            );
+            // Clear the slot signatures from status cache for this slot
+            root_bank.clear_slot_signatures(*slot);
+
+            // Clear the accounts for this slot
+            root_bank.remove_unrooted_slot(*slot);
+
+            // Clear the slot-related data in blockstore. This will:
+            // 1) Clear old shreds allowing new ones to be inserted
+            // 2) Clear the "dead" flag allowing ReplayStage to start replaying
+            // this slot
+            blockstore.clear_unconfirmed_slot(*slot);
+
+            // Signal ReplayStage to clear its progress map so that a different
+            // version of this slot can be replayed
+            let _ = duplicate_slots_reset_sender.send(*slot);
+
+            // Mark this slot as special repair, try to download from single
+            // validator to avoid corruption
+            let repair_addr = serve_repair
+                .repair_request_duplicate_compute_best_peer(*slot, cluster_slots)
+                .ok();
+            let new_duplicate_slot_repair_status = DuplicateSlotRepairStatus {
+                start: timestamp(),
+                repair_addr,
+            };
+            duplicate_slot_repair_statuses.insert(*slot, new_duplicate_slot_repair_status);
+        }
+    }
+
+    fn find_new_duplicate_slots(
+        duplicate_slot_repair_statuses: &HashMap<Slot, DuplicateSlotRepairStatus>,
+        blockstore: &Blockstore,
+        cluster_slots: &ClusterSlots,
+        root_bank: &Bank,
+    ) -> Vec<Slot> {
+        let dead_slots_iter = blockstore
+            .dead_slots_iterator(root_bank.slot() + 1)
```

Comment:
assuming currently `dead_slots_iterator` is ascending order, how about doing `Iterator::rev()` or `Vec<Slot>::reverse()`? Did you try already?

I hope that allow us to avoid [this](https://github.com/solana-labs/solana/pull/9805/files/70a2dd47c0cb8e7d9d20d4d29928643115b5b880#diff-5984b6b0429f857c13a2a362669ab37cR518) and [that](https://github.com/solana-labs/solana/pull/9805/files/70a2dd47c0cb8e7d9d20d4d29928643115b5b880#diff-a7549f152920d85fb44e6a784b8e5e1fR2273).